ENH: Quaternion SLERP

adbugger opened the pull request:
Initial implementation of quaternion slerp. I'm sure that there are a few more error checks possible, but how is the overall algorithm and implementation?

adbugger committed:
Initial commit for Rotation class

adbugger committed:
Modified docstring

adbugger committed:
Initial commit for overall package documentation

adbugger committed:
Corrections transform package main documentation and imports in __init__.py

adbugger committed:
Implemented from_quaternion. Class method not being documented.

adbugger committed:
Added tests for from_quaternion function

adbugger committed:
Shifted normalisation to `__init__` function

adbugger committed:
Added `normalized` flag. Class method still not documented.

adbugger committed:
Finally fixed documentation. All methods are now documented.

adbugger committed:
Reduced number of operations when normalizing. Added tests.

adbugger committed:
Added `transform/setup.py`, modified input shape check for clarity.

adbugger committed:
Minor change to if-else logic

adbugger committed:
STY: PEP8 compliance for `rotation.py` and `setup.py`

adbugger committed:
STY: Added actual argument shape in ValueError

adbugger committed:
MAINT: Zero norm quaternions now raise ValueError instead of UserWarning

adbugger committed:
STY: fixed whitespace in test_rotation.py

adbugger committed:
ENH: Added Rotation.as_quaternion method

The shape of the returned array depends on the initialization, and
whether the Rotation object contains a single quaternion.

adbugger committed:
TST: Tests for `Rotation.as_quaternion()` function.

Modified tests to work with the as_quaternion() function. Also
added a new test case to ensure that input of from_quaternion()
output of as_quaternion() always match in shape.

adbugger committed:
DOC: Added `as_quaternion` to methods section

adbugger committed:
ENH: Implemented as_dcm method.

adbugger committed:
TST: Added test cases for as_dcm

adbugger committed:
MAINT: Simpler and more efficient dcm component assignment

According to the tests, array indices during assignment are now
required to be transpose of the earlier intuition.

adbugger committed:
STY: Changed description of shape in docstrings

adbugger committed:
TST: Modify tests to use numpy.assert_equal instead of python assert

adbugger committed:
ENH: Added scalar implementation of from_dcm for testing only

adbugger committed:
MAINT: Updated experimental implementation

Forgot to git add to the last commit

adbugger committed:
ENH: Experimental implementation of vectorized approach

adbugger committed:
MAINT: Fixed array indexing

adbugger committed:
MAINT: Faster vector implementation.

The benchmarks on the wiki has been updated with the latest timings
and the vector implementation performs much better than the scalar
one. Moving on to cleaner implementation, documentation, and testing.

adbugger committed:
MAINT : Cleaned up code and decided on vector implementation.

adbugger committed:
TST: Added tests for from_dcm function

adbugger committed:
TST: Updated test to remove use of np.matmul

The matmul function was not introduced in numpy v1.8.2 and causes
travis-ci builds to fail. Instead of checking product of matrix and
transpose with identity, now we check equality of matrix inverse and
its transpose.

adbugger committed:
MAINT: normalized=True and np.empty in from_dcm

The `from_dcm` function now passes normalized quaternions to `__init__`
and uses np.empty to initialize `decision_matrix`.

adbugger committed:
TST: Fixed random state and random seed in tests

adbugger committed:
DOC: Updated documentation of from_dcm and added reference

adbugger committed:
TST: Used np.einsum instead of matmul

Travis-CI has been setup to test on a build with numpy==1.8.2, which
did not have np.matmul. To ensure test cases pass, Einstein summation
convention has been used to multiply matrices.

nmayorov committed:
MAINT: Normalize quat in a single place in Rotation.from_dcm

adbugger committed:
ENH: Implemented initialization from rotation vector

adbugger committed:
TST: Added test cases for `from_rotvec` function

adbugger committed:
STY: Fixed indentation for pep8

adbugger committed:
MAINT: Fixed implementation of from_rotvec

adbugger committed:
TST: Added identity quaternion in tests for `from_rotvec`

adbugger committed:
DOC: Added `from_rotvec` to documentation

adbugger committed:
MAINT: Fixed typo in function

adbugger committed:
MAINT: Renamed variable and used ** instead of np.power

adbugger committed:
MAINT: Changed implementation for clarity

Introduce `scale` variable in `from_rotvec` for clarity

adbugger committed:
TST: Added new tests and modified old ones for from_rotvec

Added a test case for small angles. Modified previous tests to use
as_quaternion function instead of accessing ._quat directly

adbugger committed:
ENH: Added implementation of `as_rotvec` function

adbugger committed:
STY: Removed redefinition of `small_norms` variable

adbugger committed:
STY: Style changes to code

adbugger committed:
TST: Better testing for small angles

adbugger committed:
STY: Removed extra blank line

adbugger committed:
TST: Shortened numbers in tests

adbugger committed:
MAINT: Used boolean masks instead of np.nonzero

adbugger committed:
TST: Added match parameter to pytest.raises(ValueError)

adbugger committed:
STY: Fixed whitespace

adbugger committed:
TST: Added test cases for `as_rotvec` function

adbugger committed:
ENH : Initial implementation of `from_euler`

Issues still need to be worked out in input format validation.
Implementation is still returning nans.

adbugger committed:
ENH: Implementation of `from_euler` function

Fixed errors and made a few decisions. Now, in the case of a single
rotations, the user can specify upto 3 angles and axes. In the case
of multiple rotations, however, numpy.asarray requires that the
dimensions match, so there the user must input a (N, 3) matrix only.

adbugger committed:
MAINT: Updated str-language specification

adbugger committed:
ENH: `from_euler` implementation using elementary quaternions

Introduced standalone functions for quaternion composition.

adbugger committed:
TST: Simple tests for `from_euler`

Initial tests. More to be added.

adbugger committed:
STY: Minor change to test file

adbugger committed:
TST: Changed test to compare quaternions. All tests pass now.

adbugger committed:
MAINT: Changed implementation of `from_euler`.

Implementation of `from_euler` has been changed according to
suggestions. Standalone functions, where appropriate, have been moved
out of the class.

adbugger committed:
TST: Add test case to check order of rotations

Intrinsic rotations should be same as extrinsic with the order of axes
reversed

adbugger committed:
MAINT: Minor changes to calculations and condition checks

adbugger committed:
TST: Elementary test to check extrinsic rotation implementation

adbugger committed:
MAINT: Changes to `from_euler`

adbugger committed:
TST: Add thorough testing for `from_euler`

adbugger committed:
STY/DOC: Rename constane, combine reference in docstring

adbugger committed:
ENH: Initial implementation of `as_euler` function. Still buggy.

adbugger committed:
MAINT: Fixed bugs with `as_euler`. Initial implementation now works.

Had missed and arccos function previously. Understood adaptations to
algorithm. Instead of reversing the sign of every angle, we need to
reverse the axis sequence and angle sequence for extrinsic rotations.

adbugger committed:
TST: Tests for `as_euler` function

adbugger committed:
DOC: Added references

The reference links are behind a paywall. I couldn't find a link to the
full texts.

adbugger committed:
STY: Style changes to code

Removed leading underscores from global function names. Renamed
argument. Calculated sin(lambda) and cos(lambda) explicitly. Removed
whitespace from docstring. Modified local variable in test.

adbugger committed:
STY: Explanation in comment is clearer

adbugger committed:
TST: Test cases for `as_euler` with extrinsic rotations

These test cases compare the actual angles returned by the algorithm. It appears that we cannot always guarantee that these will be th same as the input angles, only that the final rotation generated by the angles will be the same.

For intrinsic rotations, it appears that the algorithm never gives the same angles.

adbugger committed:
MAINT: Used np.einsum instead of loop

adbugger committed:
TST: Changed variable names in test

adbugger committed:
MAINT/DOC: Corrected einsum broadcasting. Added function to docs.

adbugger committed:
DOC: Removed redundant reference. Cleared up convention difference.

adbugger committed:
MAINT: Enforced angle restrictions

adbugger committed:
MAINT: Updated range restrictions for second angle

adbugger committed:
TST: Use randomly generated angles for testing.

Now that the ranges for angle2 have been properly fixed, we can test for
angle equality with randomly generated angles within those ranges.

adbugger committed:
TST: Split tests according to axes type, symmetric and asymmetric

adbugger committed:
MAINT: Renamed variables appropriately

adbugger committed:
DOC: Clarified angle ranges in the docstring

adbugger committed:
STY/DOC: Style changes to code, fixed error in docstring

adbugger committed:
TST: Add tests to ensure proper handling of gimbal lock

adbugger committed:
MAINT: Fixed detection of gimbal lock

The problem was that the value of eps had been taken to be too small. It
is now the smallest value that passes the tests. (~1e-7). Also make some
minor style and doc changes.

adbugger committed:
DOC: Mentioned behaviour for cases of gimbal lock in docstring

adbugger committed:
TST: Tests for gimbal lock cases

adbugger committed:
MAINT: Don't apply angle adjustment for gimbal locked cases

adbugger committed:
TST: Gimbal lock can only be caused by second angle being close to 90

adbugger committed:
DOC: Removed explanation, added reference for gimbal lock

adbugger committed:
STY: Minor changes to code style

adbugger committed:
MAINT: Handling Gimbal lock.

Clarified warning for gimbal lock. Also tried not to reverse angle
sequences for gimbal locked cases but that leads to test failures.

adbugger committed:
MAINT: Handle extrinsic rotations in a separate branch

This is to ensure that after reversing angle sequence we still get third
angle as zero

adbugger committed:
TST: Angles specific to gimbal lock cases

adbugger committed:
MAINT: Remove dcm transposition and adapt rest of algorithm

adbugger committed:
DOC: Fixed docstring

adbugger committed:
MAINT: Add leading '_' to private functions

adbugger committed:
ENH: Implementation of `inv` function

adbugger committed:
TST: Test for correct inversion of rotation

adbugger committed:
MAINT: `inv()` now respects `_single` flag

adbugger committed:
TST: Tests for single rotation inversion

adbugger committed:
MAINT: Fixed error with `_single` flag

adbugger committed:
MAINT: Concise implementation

adbugger committed:
Merge pull request #13 from adbugger/rotation_inv

Initial implementation of the inverse function

adbugger committed:
ENH: Initial implementation of rotation composition

adbugger committed:
MAINT: Simpler implementation

adbugger committed:
MAINT: Removed unused function

Since we're not using the matrix representation anymore, remove the
function responsible for generating that representation.

adbugger committed:
MAINT: Added error message for shape mismatch.

adbugger committed:
Merge pull request #14 from adbugger/rotation_mul

ENH: Initial implementation of rotation composition

adbugger committed:
DOC: Documentation changes for rotation composition function

adbugger committed:
DOC: Add __mul__ to exclusion list in class.rst

adbugger committed:
ENH: Initial implementation of `apply` function

adbugger committed:
MAINT/DOC: Explain rotation-projection duality, add inverse flag

adbugger committed:
ENH: Initial commit for `Slerp` class

adbugger committed:
DOC: Changed Rotation class docstring.

Since the interpolators and estimators are in separate classes, changed
the docstring to reflect these.

adbugger committed:
DOC: Added Slerp class to package documentation

adbugger committed:
STY: Fixed Slerp class documentation

adbugger committed:
ENH: Added the slerp interpolation method

nmayorov commented on scipy/spatial/transform/__init__.py:
```diff
@@ -19,7 +19,7 @@
 It also implements some useful algorithms such as
 
-* Quaternion slerp
+* Quaternion `Slerp` (Spherical Linear intERPolation)
```
I think don't need to emphasise ERP.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+
+    Parameters
+    ----------
+    rotations : a `Rotation` instance
```
I think `timestamps` should go before `rotations` as in (x, y) order.

nmayorov replied:
Add that `rotations` should contains (N,) rotations as well and at least two of them.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+        # Possibly sort timestamps, or add a sorted flag?
+        self._times = np.asarray(timestamps)
+
+        if self._times.shape[0] != self._rots._quat.shape[0]:
```
I suppose we need some property to get the number of rotations in `rotation`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+    __call__
+    """
+    def __init__(self, rotations, timestamps):
+        self._rots = rotations
```
OK, definitely we need to do more preparations here. 

I suppose that you should compute rotations between "keyframes" here, possible extract rotation vectors from them and store.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+    def __init__(self, rotations, timestamps):
+        self._rots = rotations
+        # Possibly sort timestamps, or add a sorted flag?
+        self._times = np.asarray(timestamps)
```
We can require that `timestamps` should be strictly increasing. While decreasing order makes some sense, I suggest not to implement it now (from my experience I know that it creates quite a bit of troubles).

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+
+        References
+        ----------
+        .. [1] `Quaternion Slerp
```
Definitely goes to `__init__`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+                <https://en.wikipedia.org/wiki/Slerp#Quaternion_Slerp>`_
+        """
+        # Possibly sort timestamps, or add a sorted flag?
+        frames = np.asarray(timestamps)
```
I think you should use `np.searchsorted`.

Basically you need to find an interval for each time from `timestamps`, find ratio in [0, 1] of the interval coverage and then do some rotation composition using prepared rotation vectors. Should be pretty straightforward.

If time is outside the whole span from `__init__`, then raise an error.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+        q = _compose_quat(q0, q1)
+        # Make the real part of q positive so that we have a rotation
+        # the "short" way
+        q[q[:, -1] < 0] *= -1
```
You don't need to do that, because we always extract the rotation vector "the short way".

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+
+        Parameters
+        ----------
+        timestamps : array_like, 1D
```
I suggest something lighter, like `t`.

Could go with `Slerp(t, r)`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,98 @@ def __mul__(self, other):
+                             ))
+
+    def __call__(self, timestamps):
+        """Generates the interpolated rotation at given times.
```
Use imperatives, "Generate ...". And generally try to improve formulations as you go through them. I mean read it, if you feel that something is not quite right with it --- try to rethink and rephrase that.

nmayorov reviewed (commented):
Some general guidelines.

adbugger committed:
DOC: Better explanation of rotation-projection duality

adbugger committed:
STY: Changed emphasis on full form of abbreviation

adbugger committed:
MAINT/DOC: Revised documentation, argument names. Used np.searchsorted

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+    orientations, and results in rotation around a single axis with
+    constant velocity.
+
+    Given a set of `rotations` and corresponding fixed `timestamps`, this class
```
I think this paragraphs is almost unnecessary.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+class Slerp(object):
+    """Spherical Linear Interpolation of Rotations.
+
+    The interpolation is done using the Spherical Linear intERPolation
```
I think you don't need to repeat "Spherical ..." again.

Write something short, like "The interpolation between consecutive rotations is performed as a rotation around a fixed axis with a constant angular rate [1]_."

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+    Parameters
+    ----------
+    t : array_like, shape (N,)
+        An array of strictly increasing timestamps for the fixed rotations. The
```
Don't need  to mention `rotations` here. 

And why being so sloppy with naming? If you rename an argument, check that you renamed it everywhere.

So keep it simple, as "Array of timestamps in strictly increasing order with at least 2 elements."

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+        An array of strictly increasing timestamps for the fixed rotations. The
+        number of times `N` must be equal to the number of rotations specified
+        in the `rotations` object.
+    r : a `Rotation` instance
```
I think it's better to avoid articles in such places, so `Rotation instance` or `Rotation`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+        number of times `N` must be equal to the number of rotations specified
+        in the `rotations` object.
+    r : a `Rotation` instance
+        This object contains the fixed rotations between which the
```
`Contains N rotations to do the interpolation between.` (or something like that).

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+            <https://en.wikipedia.org/wiki/Slerp#Quaternion_Slerp>`_
+    """
+    def __init__(self, t, r):
+        self._key_quats = r._quat
```
Why do you use private attributes so much? Use `r.as_quaternion()` if you need.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+    """
+    def __init__(self, t, r):
+        self._key_quats = r._quat
+        self._num_rots = r._quat.shape[0]
```
To be replaced with some public property. LIke `r.num_rots`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+                                self._times.shape[0],
+                             ))
+
+        # Make a copy for modification
```
I have a nice idea (or so it seems):

Let's introduce two methods `__getitem__` and `copy`. This way we could write `q0 = r[:-1].copy()`. The implementation is very straightforward.

This way we won't need to work with quaternions explicitly.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+        self._interp_rotvecs = Rotation.from_quaternion(
+            _compose_quat(q0, q1)).as_rotvec()
+
+    def __call__(self, timestamps):
```
timestamps -> t

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+            _compose_quat(q0, q1)).as_rotvec()
+
+    def __call__(self, timestamps):
+        """Generate interpolated rotations at given times.
```
I think "Compute" is a more standard verb.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+    def __call__(self, timestamps):
+        """Generate interpolated rotations at given times.
+
+        Given an array of `timestamps` returns a `Rotation` instance with the
```
This paragraph doesn't seem necessary.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+        Parameters
+        ----------
+        timestamps : array_like, 1D
+            An array of increasing timestamps, for the interpolated rotations
```
Don't think they need to be in increasing order. Think of better and shorter phrasing too.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+        timestamps : array_like, 1D
+            An array of increasing timestamps, for the interpolated rotations
+        """
+        frames = np.asarray(timestamps)
```
Why frames?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+                                self._num_rots,
+                                self._times.shape[0],
+                             ))
+
```
Where is the check of increasing order? Compute something like `dt = np.diff(t)`, check it and save for future usage.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+
+        # Index of largest keyframe time less than or equal to given time
+        # Given timestamps need not be sorted anymore
+        q0_ind = np.searchsorted(self._times, frames, side='right') - 1
```
`side='right'` isn't necessary, correct?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+                             "keyframe timestamps [{}, {}].".format(
+                                self._times[0], self._times[-1]))
+
+        q0 = self._key_quats[q0_ind]
```
I don't know why, but it all reads very awkward. Try different names, with less `_` for example.

Like I don't think it will be less readable if you name it like `q0_ind` -> `ind`, `q0_q1_rotvec -> rotvec`, etc.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +656,106 @@ def __mul__(self, other):
+
+        q0 = self._key_quats[q0_ind]
+        q0_q1_rotvec = self._interp_rotvecs[q0_ind]
+        t = ((frames - self._times[q0_ind]) /
```
Name it `alpha` or something.

nmayorov reviewed (commented):
I suggested two new methods, which should make implementing this class clearn.

Work on the tests.

adbugger committed:
ENH: Add number of rotations as a public property to Rotation class

adbugger committed:
ENH: Add __getitem__ and copy methods for indexing and slicing

adbugger committed:
MAINT/DOC: Updated documentation and implementation of Slerp.__init__

Documentation is clearer and more concise. Variable names have been
simplified. Function now deals with rotations directly instead of
working with quaternions.

adbugger committed:
MAINT: Removed type checks to support rotation reordering, given indexes

adbugger committed:
STY: Renamed variables, changed comments

adbugger committed:
MAINT/DOC: Updates to Slerp().__call__

Clearer docstrings, comments and variable names. Implementation deals
directly with rotations except where exponentiation of quaternions is
necessary.

adbugger committed:
BUG: Fixed bugs related to number of rotations

adbugger commented:
During implementation, I realized that the `copy` method is not required for interpolations, since `__getitem__` and `inv()` return new objects anyway. For now I haven't added `copy` to the documentation as we might decide to remove it altogether.

Also, should the `__getitem__` method be documented?

adbugger committed:
TST: Test cases for quaternion slerp

adbugger committed:
STY: Changed error message

adbugger committed:
TST: Check for error cases

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
             result = result[0]
         return self.__class__(result, normalized=True)
+
+    def __getitem__(self, val):
```
I guess val->index or ind.

adbugger replied:
This is just to emphasize that we can have an index, a slice, or an index array.

nmayorov replied:
I suggest `indexer` then --- something than can index an array.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+        # Returns a single rotation for an index, and a stack of rotation(s)
+        # for slice, index array, etc even if only one index is ultimately
+        # given
+        return self.__class__.from_quaternion(self._quat[val], normalized=True)
```
Why not self.__class__()?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+        return self.__class__.from_quaternion(self._quat[val], normalized=True)
+
+    def copy(self):
+        return self.__class__.from_quaternion(
```
I think you might be mistaken that `__init__` actually does copying, when `normalized=True`.

adbugger replied:
I'll look into it.

adbugger replied:
You're right, `__init__` doesn't actually copy when `normalized==True`. If any modifications are made to the input quaternion array later on in the code, they are reflected in the `self._quat` property as well. Should we be copying the quaternions in `__init__` itself to ensure such things cannot happen? This would lead to `__getitem__` also always returning new objects.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+    t : array_like, shape (N,)
+        Times of the known rotations. At least 2 times must be specified.
+    r : `Rotation` instance
+        `N` rotations to perform the interpolation between. At least 2
```
Well I believe `N` is formally not an argument, so you shouldn't write it as `N`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -224,6 +223,7 @@ def __init__(self, quat, normalized=False):
             # Normalize each quaternion, ensuring norm is broadcasted along
             # each column.
             self._quat[~zero_norms] /= norms[~zero_norms][:, None]
+        self.num_rots = self._quat.shape[0]
```
I suggest to make it property without setter.

adbugger replied:
Can you explain what you mean by that? A property without setter?

nmayorov replied:
I believe you should write 

```
@property
def num_rots():
  return self._quat.shape[0]
```
I would probably suggest to name it as just `n`, like `rot.n` --- sounds find, another option `n_rotations` (scikit-learn style). Don't really like `num_rots`.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+        if np.any(self.timedelta <= 0):
+            raise ValueError("Times must be in strictly increasing order.")
+
+        # inv() and __getitem__ return new objects anyway
```
Agree with inv(), really not sure with __getitem__. Check this carefully.

adbugger replied:
You're right, `__getitem__` does not return a new object (since `__init__` does not copy when `normalized=True`). Should I call copy here only, or we can always copy the quaternion matrix in `__init__` itself? (Refer to my comment above; copying in `__init__` would remove another source of potential errors in user code)

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+                             "number of timestamps given, got {} rotations "
+                             "and {} timestamps.".format(
+                                r.num_rots, t.shape[0]))
+        self.times = t
```
I'm not sure here, but it seems a bit illogical to use short names for API arguments and longer names inside. Perhaps use `times`, `rotation` everywhere, or stick to `t`, `dt`, etc.

I think I would prefer to have `times, rotations` as arguments (no `timestamps` though 😄 )

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+        self.rots = r[:-1]
+        # should be equal to r.rots - 1
+        self.num_rots = self.rots.num_rots
+        rot_params = self.rots.inv() * r[1:]
```
This name is hard to understand. Perhaps the best idea is not to introduce any intermediate variables.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+        # of the time range. 'left' excludes t_min.
+        ind = np.searchsorted(self.times, t) - 1
+        # Include t_min. Without this step, index for t_min equals -1
+        ind[t == self.times[0]] = 0
```
If `right` will eliminate that, then use it. Sorry if I wasn't too attentive here. Use your best judgement.

A lot of comments is usually a bad sign. Implement it as simple as possible.

adbugger replied:
Both `left` and `right` exclude one endpoint of the time range. I don't see a simpler way of implementing this.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+                                self.times[0], self.times[-1]))
+
+        rots = self.rots[ind]
+        interp_exponents = (t - self.times[ind]) / self.timedelta[ind]
```
What `exponents` mean here?

adbugger replied:
In the slerp formula, this parameter is the exponent of  `q0^(-1) * q1`. I'll change this name.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
             result = result[0]
         return self.__class__(result, normalized=True)
+
+    def __getitem__(self, val):
```
Definitely needs to be documented. You should explain that only one index should be given and what it does.

adbugger replied:
But we're not using a single index. In my code, I'm using it for slicing and reordering.

nmayorov replied:
You can write `r[1:2, 3]` --- but this would be illegal for our application. In this case you will receive a tuple with slice and int.  This is what I meant, that a single indexing object is required. I suggest not to check for such kind of errors on our side, but explain it in the documentation, that you should select rotations using "one" indexer.

adbugger replied:
Ah, ok.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,105 @@ def __mul__(self, other):
+        # should be equal to r.rots - 1
+        self.num_rots = self.rots.num_rots
+        rot_params = self.rots.inv() * r[1:]
+        self.interp_rotvecs = rot_params.as_rotvec()
```
In my opinion stacking two and more abbreviate names reads awkward. I don't know, maybe you can improve the naming somehow.

nmayorov reviewed (commented):
Didn't read tests yet.

Looks much better. But check this issue with copying, indexing and slicing.

adbugger committed:
DOC: Add documentation for copy and __getitem__ methods

adbugger committed:
MAINT: `__init__` and `as_quaternion` copy quaternions

The logic is that, once initialized, the user should not be able to
modify the quaternion array stored inside the object.

adbugger committed:
STY: Renamed variables

adbugger commented:
So I decided to have `__init__` copy the quaternions in both cases. This ensures consistent behaviour and the user cannot modify the quaternions accidentally by:
```python
r = Rotation.from_quaternion(q)
# Some code
q[:, -1] *= -1
```
Previously, this would have modified the quaternions in the object.

Further `as_quaternion` also copies the quaternions before returning them, to address the same concern. The user should not be able to modify object internals, accidentally or otherwise.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+        Parameters
+        ----------
+        val : index, slice, or index array
+            This parameter specifies which rotation(s) to extract.
```
I think you can write similar descriptions as `Specifies which rotation(s) to extract.".

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+    def __getitem__(self, val):
+        """Extract rotation at given index(es) from object.
+
+        This function returns a new `Rotation` instance containing:
```
I don't know, I think you can just say that semantics is the same as for numpy arrays and Python lists.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -209,10 +210,8 @@ def __init__(self, quat, normalized=False):
-            self._quat = quat
-        else:
-            self._quat = quat.copy()
+        self._quat = quat.copy()
```
It should have a flag `copy=True` (or `False`?). We often create a quaternion from computations (from_dcm, from_rotvec), so we don't need to make another copy here.

adbugger replied:
I suggest `copy=True` by default. For `normalized=False` we always copy, and we want to minimize user errors when `normalized=True` so I think we should copy the quaternions by default unless otherwise specified.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+        # __init__ now copies for all cases
+        return self.__class__(self._quat[val], normalized=True)
+
+    def copy(self):
```
Do we need this function after you added copying in the constructor?

adbugger replied:
No we don't need it for our code. Should I leave it as a function for the user or remove it?

nmayorov replied:
Considering our object is now essentially immutable, I don't see it being necessary at all.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+    times : array_like, shape (N,)
+        Times of the known rotations. At least 2 times must be specified.
+    rotations : `Rotation` instance
+        `N` rotations to perform the interpolation between. At least 2
```
I suggest to rewrite is as `Rotations to perform the interpolation between. Must contain N rotations.`

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+            <https://en.wikipedia.org/wiki/Slerp#Quaternion_Slerp>`_
+    """
+    def __init__(self, times, rotations):
+        # We don't need to store all rotations, just rotations[:-1] and
```
Why this comment is hanging here?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+            raise ValueError("Times must be in strictly increasing order.")
+
+        # inv() and __getitem__ return new objects anyway
+        self.rots = rotations[:-1]
```
rots -> rotations?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+
+        alpha = (compute_times - self.times[ind]) / self.timedelta[ind]
+
+        # Ensure the 1D alpha array is broadcasted column wise
```
I think this is kinda obvious, I suggest to remove this comment.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+        # inv() and __getitem__ return new objects anyway
+        self.rots = rotations[:-1]
+        # should be equal to r.rots - 1
+        self.num_rots = self.rots.num_rots
```
You seem to use it only in one place in the code, I suggest to remove it.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+        # interpolation rotvecs
+        if rotations.num_rots == 1:
+            raise ValueError("Expected at least 2 rotations in object, "
+                             "got {}.".format(rotations.num_rots))
```
I suppose got 1 ) I guess you don't need to say that.

So "`rotations` must contain at least 2 rotations.`

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -657,3 +657,119 @@ def __mul__(self, other):
+
+    Methods
+    -------
+    __call__
```
Add `__getitem__`?

adbugger replied:
`__getitem__` for slerp should return what exactly? The `ith` keyframe rotation?

nmayorov replied:
Apologies, I thought it was related to Rotation class.

nmayorov replied:
Apologies, I thought it was related to Rotation class.

nmayorov reviewed (commented):
I think it is almost ready. Didn't check the tests yet.

adbugger committed:
DOC: Explained __getitem__ semantics

adbugger committed:
ENH: Added `copy` flag in `__init__`

adbugger committed:
MAINT: Removed copy function

adbugger committed:
ENH: Added number of rotations as a property without setter

adbugger committed:
DOC: Change to slerp docstring

adbugger committed:
ENH: Use `copy=False` when initializing from computed quaternions

adbugger committed:
DOC: Docstring change for __getitem__

adbugger committed:
DOC: Docstring for property

adbugger committed:
ENH: dot semantics for apply

adbugger committed:
DOC: Shorter explanation of output cases

adbugger committed:
TST: Initial tests for rotation application

adbugger committed:
ENH: Implementation of __getitem__ function

adbugger committed:
DOC: Add __getitem__ to exclusion list

adbugger committed:
ENH: Add copy functionality to __init__

adbugger committed:
ENH: from_quaternion now uses copy flag

adbugger committed:
ENH: from_dcm uses copy flag

adbugger committed:
ENH: from rotvec uses copy flag

adbugger committed:
ENH: from_euler uses copy flag

adbugger committed:
ENH: inv uses copy flag

adbugger committed:
ENH: __mul__ uses copy flag

adbugger committed:
ENH: Add number of rotation as a property without setter

adbugger committed:
MAINT: Apply function uses num_rotations property

adbugger committed:
DOC: Small changes to docstrings.

adbugger committed:
TST: Tests use np.expand_dims function

adbugger committed:
TST: Add test with inverse=True

adbugger committed:
DOC/STY: Minor docstring and style changes

adbugger committed:
TST: Add tests for indexing and get item

adbugger committed:
Merge pull request #15 from adbugger/rotation_apply

ENH: Initial implementation of `apply` function

adbugger committed:
TST: Update previous tests to resolve merge conflicts

adbugger committed:
TST: Update previous tests to resolve merge conflicts

adbugger committed:
Merge branch 'rotation' into rotation_index_copy

adbugger committed:
Merge branch 'rotation_index_copy' of https://github.com/adbugger/scipy into rotation_index_copy

adbugger commented:
Will resolve conflicts after #17 is merged.

adbugger committed:
MAINT: Remove copy flag from `from_quaternion`

adbugger committed:
Merge pull request #17 from adbugger/rotation_index_copy

ENH: Indexing, copying and addition of number property to rotation class

adbugger committed:
MAINT: Explicit imports in `__init__.py`

adbugger committed:
MAINT: as_quaternion returns a copy of quaternion array

adbugger committed:
MAINT: Replace property `n` with `len` function

adbugger committed:
TST: Test to check against accidental corruption of object

adbugger committed:
DOC: Rotation class docstring changes.

Listed representations and operations on rotaitons.

adbugger committed:
DOC: Changes to from_quaternion docstring

Added a returns section along with reference to quaternions for spatial
rotation.

adbugger committed:
DOC: Add returns section and references to as_quaternion

adbugger committed:
DOC: Returns section, docsting imrovements for ..._dcm functions

adbugger committed:
DOC: Remove reference to class internals

adbugger committed:
DOC: Returns section and docstring improvements for rotvec methods

adbugger committed:
DOC: Better documentation for euler methods

adbugger committed:
DOC: Documentation for inverse and composition

adbugger committed:
DOC: Improved documentation for rotation application

adbugger committed:
DOC: Documenting rotation indexing

adbugger committed:
DOC: Documentation for len function

adbugger committed:
Merge branch 'rotation' into rotation_slerp

adbugger committed:
STY: Style changes after fixing merge conflicts

adbugger committed:
BUG: Added Slerp class to explicit import

adbugger committed:
MAINT: Replaced property n with len() in Slerp and test

adbugger committed:
DOC: Reorder method list. Remove section from package docstring.

adbugger committed:
STY: Reorder imports, remove unhelpful comments

adbugger committed:
STY: Reorder functions in code. Rename ..._quaternion to ..._quat.

adbugger committed:
DOC: Changed function docstrings.

More specific output names. Added references for rotvecs and dcms.
Simplified two-to-one mapping explanation in as_quat.

adbugger committed:
BUG: Fix nan error in degenerate euler axes

For certain gimbal locked cases, it appears that certain entries in
`dcm_transformed` are greater than unit norm. That results in nan values
when arccos function us used.

adbugger committed:
STY: Style changes to code

nmayorov commented:
Well, I think it is ready to be merged. However, should we wait before the Rotation is merged into scipy?

adbugger committed:
DOC: Documentation improvements

adbugger committed:
STY: Use len instead of shape[0]

adbugger committed:
Merge branch 'rotation' into rotation_slerp

adbugger commented:
If the tests pass, I think we can include this in the main PR?

adbugger committed:
TST: Use as and from_quat in tests

nmayorov commented:
I suggest to merge it first. Let new pull requests hang here for a while.

By the way, in general you should prefer rebasing over merging (at least this is how it's done for github workflow). So you just apply old commits over new branch state.

adbugger commented:
> By the way, in general you should prefer rebasing over merging (at least this is how it's done for github workflow). So you just apply old commits over new branch state.

I tried that, but it didn't work in this case due to merge conflicts. The merge commit is a result of resolving the conflicts using the online editor.
